src/main/scala/is/hail/expr/ir/SpecializedArrayBuilders.scala

```diff
@@ -0,0 +1,236 @@
+
+  def size: Int = size_
+
+  def apply(i: Int): Byte = {
+    require(i >= 0 && i < size)
+    b(i)
+  }
+
+  def ensureCapacity(n: Int) {
+    if (b.length < n) {
+      val newCapacity = (b.length * 2).max(n)
+      val newb = new Array[Byte](newCapacity)
+      Array.copy(b, 0, newb, 0, size_)
+      b = newb
+    }
+  }
+
+  def add(x: Byte) {
+    ensureCapacity(size_ + 1)
+    b(size_) = x
+    size_ += 1
+  }
+}
+
+class IntArrayBuilder(initialCapacity: Int) {
+  private var b: Array[Int] = new Array[Int](initialCapacity)
+  private var size_ : Int = 0
+
+  def size: Int = size_
+
+  def apply(i: Int): Int = {
+    require(i >= 0 && i < size)
+    b(i)
+  }
+
+  def ensureCapacity(n: Int) {
+    if (b.length < n) {
+      val newCapacity = (b.length * 2).max(n)
+      val newb = new Array[Int](newCapacity)
+      Array.copy(b, 0, newb, 0, size_)
+      b = newb
+    }
+  }
+
+  def add(x: Int) {
+    ensureCapacity(size_ + 1)
+    b(size_) = x
+    size_ += 1
+  }
+
+  def clear() {  size_ = 0 }
+}
+
+class LongArrayBuilder(initialCapacity: Int) {
+  private var b: Array[Long] = new Array[Long](initialCapacity)
+  private var size_ : Int = 0
+
+  def size: Int = size_
+
+  def apply(i: Int): Long = {
+    require(i >= 0 && i < size)
+    b(i)
+  }
+
+  def ensureCapacity(n: Int) {
+    if (b.length < n) {
+      val newCapacity = (b.length * 2).max(n)
+      val newb = new Array[Long](newCapacity)
+      Array.copy(b, 0, newb, 0, size_)
+      b = newb
+    }
+  }
+
+  def add(x: Long) {
+    ensureCapacity(size_ + 1)
+    b(size_) = x
+    size_ += 1
+  }
+
+  def clear() {  size_ = 0 }
+}
+
+class FloatArrayBuilder(initialCapacity: Int) {
+  private var b: Array[Float] = new Array[Float](initialCapacity)
+  private var size_ : Int = 0
+
+  def size: Int = size_
+
+  def apply(i: Int): Float = {
+    require(i >= 0 && i < size)
+    b(i)
+  }
+
+  def ensureCapacity(n: Int) {
+    if (b.length < n) {
+      val newCapacity = (b.length * 2).max(n)
+      val newb = new Array[Float](newCapacity)
+      Array.copy(b, 0, newb, 0, size_)
+      b = newb
+    }
+  }
+
+  def add(x: Float) {
+    ensureCapacity(size_ + 1)
+    b(size_) = x
+    size_ += 1
+  }
+
+  def clear() {  size_ = 0 }
+}
+
+class DoubleArrayBuilder(initialCapacity: Int) {
+  private var b: Array[Double] = new Array[Double](initialCapacity)
+  private var size_ : Int = 0
+
+  def size: Int = size_
+
+  def apply(i: Int): Double = {
+    require(i >= 0 && i < size)
+    b(i)
+  }
+
+  def ensureCapacity(n: Int) {
+    if (b.length < n) {
+      val newCapacity = (b.length * 2).max(n)
+      val newb = new Array[Double](newCapacity)
+      Array.copy(b, 0, newb, 0, size_)
+      b = newb
+    }
+  }
+
+  def add(x: Double) {
+    ensureCapacity(size_ + 1)
+    b(size_) = x
+    size_ += 1
+  }
+
+  def clear() {  size_ = 0 }
+}
+
+class BooleanArrayBuilder(initialCapacity: Int) {
+  private var b: Array[Boolean] = new Array[Boolean](initialCapacity)
```

Comment:
An opportunity for future performance improvement: Arrays of Boolean are typically encoded as Array of Byte by the JVM. We could store Array[Long] and mask out the bit we want.